### DISCUSSION TITLE
Loading GLTF/assets binary directly from memory

### DISCUSSION THREAD

**Discussion by vincent-herlemont** [bevyengine/bevy#13602] [Q&A]
I have a gltf binary in memory available before launching the App Bevy. There is other code that generates this binary, and I would like to avoid saving it to the disk and then loading it into Bevy.

Is there a way to create dynamicalty an embedded assets or assets directly from memory instead of going through the file system?

For the moment, I have the impression that we can only create embedded assets from a file source [bevy::asset::io::embedded::EmbeddedAssetRegistry](https://docs.rs/bevy/latest/bevy/asset/io/embedded/struct.EmbeddedAssetRegistry.html)?

What would be the best way to do it?

**Comment by hymm** [comment]
I think you can use `insert_asset` on `EmbeddedAssetRegistry`. You just need to give it a fake path, that you use to load the asset.

**Reply by vincent-herlemont** [reply]
@hymm Thank you for your response. I tried inserting the asset this way:

```rust
   // main.rs
   let mut app = App::new();
   ...
   app.add_plugins(
            DefaultPlugins
                .build()
                .disable::<LogPlugin>()
        )
        .init_resource::<SceneController>();
     
    let embedded_asset_registry = app.world_mut().resource_mut::<EmbeddedAssetRegistry>();
    embedded_asset_registry.insert_asset(PathBuf::from(""), &PathBuf::from("scene2.glb"), gltf_file_byte);
    
    app.add_systems(Startup, setup)
        .run();
    ...
```
Question: Is there a particular order that needs to be followed? 

And in the `setup` function, I tried to get the asset in several ways, but none of them worked. I was inspired by this example [embedded_asset.rs](https://github.com/bevyengine/bevy/blob/main/examples/asset/embedded_asset.rs) :

```rust
    // The name of the crate is "crate_example".

    let path = Path::new("crate_example").join("scene2.glb");
    let source = AssetSourceId::from("embedded");
    let asset_path = AssetPath::from_path(&path).with_source(source);

    commands.spawn(SceneBundle {
        scene: asset_server.load(asset_path),
        ..default()
    });

    commands.spawn(SceneBundle {
        scene: asset_server.load("scene2.glb#Scene0"),
        ..default()
    });

    commands.spawn(SceneBundle {
        scene: asset_server.load("embedded://scene2.glb#Scene0"),
        ..default()
    });

    commands.spawn(SceneBundle {
        scene: asset_server.load("embedded://crate_example/scene2.glb#Scene0"),
        ..default()
    });
```
Currently the result during loading is:

```bash
 ERROR bevy_asset::server           > Path not found: crate_example/scene2.glb
 ERROR bevy_asset::server           > Path not found: /<home_path>/assets/scene2.glb
 ERROR bevy_asset::server           > Path not found: crate_example/scene2.glb
 ERROR bevy_asset::server           > Path not found: crate_example/scene2.glb
```

The path is not founded ❌

**Reply by Fidius-jko** [reply]
Maybe create Asset Reader?

**Reply by Fidius-jko** [reply]
> @hymm Thank you for your response. I tried inserting the asset this way:
> 
> ```rust
>    // main.rs
>    let mut app = App::new();
>    ...
>    app.add_plugins(
>             DefaultPlugins
>                 .build()
>                 .disable::<LogPlugin>()
>         )
>         .init_resource::<SceneController>();
>      
>     let embedded_asset_registry = app.world_mut().resource_mut::<EmbeddedAssetRegistry>();
>     embedded_asset_registry.insert_asset(PathBuf::from(""), &PathBuf::from("scene2.glb"), gltf_file_byte);
>     
>     app.add_systems(Startup, setup)
>         .run();
>     ...
> ```
> 
> Question: Is there a particular order that needs to be followed?
> 
> And in the `setup` function, I tried to get the asset in several ways, but none of them worked. I was inspired by this example [embedded_asset.rs](https://github.com/bevyengine/bevy/blob/main/examples/asset/embedded_asset.rs) :
> 
> ```rust
>     // The name of the crate is "crate_example".
> 
>     let path = Path::new("crate_example").join("scene2.glb");
>     let source = AssetSourceId::from("embedded");
>     let asset_path = AssetPath::from_path(&path).with_source(source);
> 
>     commands.spawn(SceneBundle {
>         scene: asset_server.load(asset_path),
>         ..default()
>     });
> 
>     commands.spawn(SceneBundle {
>         scene: asset_server.load("scene2.glb#Scene0"),
>         ..default()
>     });
> 
>     commands.spawn(SceneBundle {
>         scene: asset_server.load("embedded://scene2.glb#Scene0"),
>         ..default()
>     });
> 
>     commands.spawn(SceneBundle {
>         scene: asset_server.load("embedded://crate_example/scene2.glb#Scene0"),
>         ..default()
>     });
> ```
> 
> Currently the result during loading is:
> 
> ```shell
>  ERROR bevy_asset::server           > Path not found: crate_example/scene2.glb
>  ERROR bevy_asset::server           > Path not found: /<home_path>/assets/scene2.glb
>  ERROR bevy_asset::server           > Path not found: crate_example/scene2.glb
>  ERROR bevy_asset::server           > Path not found: crate_example/scene2.glb
> ```
> 
> The path is not founded ❌

Check path in embedded_path!()

**Reply by vincent-herlemont** [reply]
> Maybe create Asset Reader?


Thank you, @Fidius-jko . What do you mean by that "Maybe create Asset Reader?"?  Why would a custom reader be needed when I want to load data from embedded, a known format like glTF?


> Check path in embedded_path!()

As for [`embedded_path!()`](https://docs.rs/bevy/latest/bevy/asset/macro.embedded_path.html), here it is: `crate_example/api/renderer/scene2.glb`. 

But why would that have an impact in my case? I am not using the [`embedded_asset!()`](https://docs.rs/bevy/latest/bevy/asset/macro.embedded_asset.html) macro, so normally the naming rule by module defined by this line [`let crate_name = module_path!().split(':').next().unwrap();`](https://github.com/bevyengine/bevy/blob/44c8cc66c4ebf3f897948dd62fa5683f32b05b91/crates/bevy_asset/src/io/embedded/mod.rs#L114) shouldn't have an impact in my case. Because I am directly using the [`insert_asset`](https://docs.rs/bevy/latest/bevy/asset/io/embedded/struct.EmbeddedAssetRegistry.html#method.insert_asset) function, I should be able to set the `asset_path` of my choice.


In other words, it seems to me that this loading should work (but it doesn't ❌ ).

```rust
    // init
    let embedded_asset_registry = app.world_mut().resource_mut::<EmbeddedAssetRegistry>();
    embedded_asset_registry.insert_asset(PathBuf::from(""), &PathBuf::from("scene2.glb"), gltf_file_byte);


    // ...
    // setup
    commands.spawn(SceneBundle {
        scene: asset_server.load("embedded://scene2.glb#Scene0"),
        ..default()
    });
```

Result:

```rust
 ERROR bevy_asset::server           > Path not found: scene2.glb
```

**Reply by Fidius-jko** [reply]
> > Maybe create Asset Reader?
> 
> Thank you, @Fidius-jko . What do you mean by that "Maybe create Asset Reader?"? Why would a custom reader be needed when I want to load data from embedded, a known format like glTF?

`AssetReader` is loading resources. Bevy have `FileAssetReader` and we load files with it by default.
We can create `AssetSourceBuilder` by `register_asset_source`. And use it.

**Comment by ThomasVille** [comment]
I agree it would be much more convenient to have a way to load an asset from a `io::Read + io::Seek` directly instead of having to provide a fake path and a custom `AssetReader`.

**Comment by ExpertOfNil** [comment]
> I have a gltf binary in memory available before launching the App Bevy. There is other code that generates this binary, and I would like to avoid saving it to the disk and then loading it into Bevy.
> 
> Is there a way to create dynamicalty an embedded assets or assets directly from memory instead of going through the file system?
> 
> For the moment, I have the impression that we can only create embedded assets from a file source [bevy::asset::io::embedded::EmbeddedAssetRegistry](https://docs.rs/bevy/latest/bevy/asset/io/embedded/struct.EmbeddedAssetRegistry.html)?
> 
> What would be the best way to do it?

Were you ever able to come up with a viable solution to this?  It seems like there should be some way to leverage the `GltfLoader` for this, but I haven't been able to figure out how.

**Reply by ExpertOfNil** [reply]
I found this gem (which uses the `MemoryAssetReader`) as I was about to give up: https://www.reddit.com/r/bevy/comments/1i83wv5/tutorial_how_to_load_inmemory_assets_in_bevy/
